File: backend/app/schemas/invest.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


class TokenIn(BaseModel):
    token: str = Field(min_length=20, max_length=4096)


class ConnectionOut(BaseModel):
    configured: bool
    token_hint: str | None = None
    status: str | None = None
    last_sync_at: datetime | None = None
    last_error: str | None = None
    history_from: date | None = None
    accounts_count: int = 0
    operations_count: int = 0
    positions_count: int = 0


class SyncRunOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    trigger: str
    status: str
    started_at: datetime
    finished_at: datetime | None
    accounts_count: int
    operations_count: int
    instruments_count: int
    positions_count: int
    error_message: str | None
    notes: str | None


class AccountOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    broker_account_id: str
    name: str
    type: str
    status: str


class PositionOut(BaseModel):
    account_id: int
    account_name: str
    figi: str
    ticker: str
    name: str
    instrument_type: str
    quantity: str
    average_price: str
    average_price_currency: str
    current_price: str
    current_price_currency: str


class DashboardPositionOut(BaseModel):
    account_id: int
    account_name: str
    figi: str
    ticker: str
    name: str
    instrument_type: str
    quantity: str
    average_price: str
    average_price_currency: str
    current_price: str
    current_price_currency: str
    value: str
    cost: str
    pnl: str
    pnl_percent: str | None = None
    share: str
    is_cash: bool
    average_source: str


class DashboardOut(BaseModel):
    value: str
    invested: str
    profit: str
    profit_percent: str | None = None
    cash: str
    prices_as_of: datetime | None = None
    prices_live: bool = False
    history_from: date | None = None
    invested_missing: bool = False
    xirr_percent: str | None = None
    xirr_from: date | None = None
    positions: list[DashboardPositionOut]


class OperationOut(BaseModel):
    id: int
    account_name: str
    broker_operation_id: str
    operation_type: str
    name: str
    figi: str
    ticker: str
    quantity: str
    payment: str
    currency: str
    occurred_at: datetime

File: frontend/src/api/invest.ts
import { apiJson } from "./http";

export type ConnectionStatus = {
  configured: boolean;
  token_hint: string | null;
  status: string | null;
  last_sync_at: string | null;
  last_error: string | null;
  history_from: string | null;
  accounts_count: number;
  operations_count: number;
  positions_count: number;
};

export type SyncRun = {
  id: number;
  trigger: string;
  status: string;
  started_at: string;
  finished_at: string | null;
  accounts_count: number;
  operations_count: number;
  instruments_count: number;
  positions_count: number;
  error_message: string | null;
  notes: string | null;
};

export type Account = {
  id: number;
  broker_account_id: string;
  name: string;
  type: string;
  status: string;
};

export type Position = {
  account_id: number;
  account_name: string;
  figi: string;
  ticker: string;
  name: string;
  instrument_type: string;
  quantity: string;
  average_price: string;
  average_price_currency: string;
  current_price: string;
  current_price_currency: string;
};

export type Operation = {
  id: number;
  account_name: string;
  broker_operation_id: string;
  operation_type: string;
  name: string;
  figi: string;
  ticker: string;
  quantity: string;
  payment: string;
  currency: string;
  occurred_at: string;
};

export function getConnection() {
  return apiJson<ConnectionStatus>("/api/settings/connection");
}

export function saveToken(token: string) {
  return apiJson<ConnectionStatus>("/api/settings/connection", {
    method: "PUT",
    body: JSON.stringify({ token }),
  });
}

export function startSync() {
  return apiJson<SyncRun>("/api/sync", { method: "POST" });
}

export function getSyncRuns() {
  return apiJson<SyncRun[]>("/api/sync/runs");
}

export function getAccounts() {
  return apiJson<Account[]>("/api/accounts");
}

export function getPositions() {
  return apiJson<Position[]>("/api/positions");
}

export function getOperations() {
  return apiJson<Operation[]>("/api/operations");
}
